File: packages/core/src/compiler.ts
import {
  CanonicalizationError,
  captureCanonicalJson,
  compareUnicodeCodePoints,
  hashCanonicalSerialization,
} from "./canonical.js";
import { validateGraphDocument } from "./schema-validation.js";
import type { EdgeSpec, GraphSpec } from "./types.js";

export type DiagnosticSeverity = "error" | "warning";

export type DiagnosticCode =
  | "GE1001_DUPLICATE_NODE"
  | "GE1002_DUPLICATE_EDGE"
  | "GE1003_MISSING_SOURCE"
  | "GE1004_MISSING_TARGET"
  | "GE1005_CYCLE"
  | "GE1006_UNREACHABLE_NODE"
  | "GE1007_INVALID_GRAPH"
  | "GE1008_MISSING_ENTRYPOINT"
  | "GE1009_MISSING_OUTPUT"
  | "GE1010_ENTRYPOINT_HAS_INCOMING"
  | "GE1101_MAX_FAN_OUT"
  | "GE1102_MAX_DEPTH";

export interface CompilerDiagnostic {
  code: DiagnosticCode;
  severity: DiagnosticSeverity;
  message: string;
  path?: string;
  nodeIds?: readonly string[];
  edgeId?: string;
}

export interface CompilationResult {
  valid: boolean;
  graphHash: string | null;
  canonicalGraph: string | null;
  entrypoints: readonly string[];
  topologicalLayers: readonly (readonly string[])[];
  diagnostics: readonly CompilerDiagnostic[];
}

function sorted(values: Iterable<string>): string[] {
  return [...values].sort(compareUnicodeCodePoints);
}

function diagnostic(
  code: DiagnosticCode,
  message: string,
  fields: Omit<CompilerDiagnostic, "code" | "severity" | "message"> = {},
): CompilerDiagnostic {
  return { code, severity: "error", message, ...fields };
}

/** Compile and statically validate a v1alpha1 graph. */
export function compileGraph(document: GraphSpec | unknown): CompilationResult {
  const diagnostics: CompilerDiagnostic[] = [];
  let captured: ReturnType<typeof captureCanonicalJson>;
  try {
    captured = captureCanonicalJson(document);
  } catch (error) {
    const message = error instanceof CanonicalizationError
      ? error.message
      : "Graph input could not be inspected safely";
    return {
      valid: false,
      graphHash: null,
      canonicalGraph: null,
      entrypoints: [],
      topologicalLayers: [],
      diagnostics: [diagnostic("GE1007_INVALID_GRAPH", message)],
    };
  }

  let schemaIssues: readonly string[];
  try {
    schemaIssues = validateGraphDocument(captured.value);
  } catch {
    return {
      valid: false,
      graphHash: null,
      canonicalGraph: null,
      entrypoints: [],
      topologicalLayers: [],
      diagnostics: [
        diagnostic("GE1007_INVALID_GRAPH", "Graph validation could not be completed safely"),
      ],
    };
  }
  if (schemaIssues.length > 0) {
    const invalid = diagnostic("GE1007_INVALID_GRAPH", schemaIssues.join("; "));
    return {
      valid: false,
      graphHash: null,
      canonicalGraph: null,
      entrypoints: [],
      topologicalLayers: [],
      diagnostics: [invalid],
    };
  }
  const graph = captured.value as unknown as GraphSpec;
  const canonicalGraph = captured.serialized;
  const graphHash = hashCanonicalSerialization(canonicalGraph);

  const nodesById = new Map<string, GraphSpec["nodes"][number]>();
  for (const [index, node] of graph.nodes.entries()) {
    if (nodesById.has(node.id)) {
      diagnostics.push(
        diagnostic("GE1001_DUPLICATE_NODE", `Node id '${node.id}' is declared more than once`, {
          path: `#/nodes/${index}/id`,
          nodeIds: [node.id],
        }),
      );
      continue;
    }
    nodesById.set(node.id, node);
  }

  const seenEdgeIds = new Set<string>();
  const validEdges: EdgeSpec[] = [];
  for (const [index, edge] of graph.edges.entries()) {
    if (seenEdgeIds.has(edge.id)) {
      diagnostics.push(
        diagnostic("GE1002_DUPLICATE_EDGE", `Edge id '${edge.id}' is declared more than once`, {
          path: `#/edges/${index}/id`,
          edgeId: edge.id,
        }),
      );
    } else {
      seenEdgeIds.add(edge.id);
    }

    const sourceExists = nodesById.has(edge.from.node);
    const targetExists = nodesById.has(edge.to.node);
    if (!sourceExists) {
      diagnostics.push(
        diagnostic("GE1003_MISSING_SOURCE", `Edge '${edge.id}' references missing source '${edge.from.node}'`, {
          path: `#/edges/${index}/from/node`,
          edgeId: edge.id,
          nodeIds: [edge.from.node],
        }),
      );
    }
    if (!targetExists) {
      diagnostics.push(
        diagnostic("GE1004_MISSING_TARGET", `Edge '${edge.id}' references missing target '${edge.to.node}'`, {
          path: `#/edges/${index}/to/node`,
          edgeId: edge.id,
          nodeIds: [edge.to.node],
        }),
      );
    }
    if (sourceExists && targetExists) {
      validEdges.push(edge);
    }
  }

  const entrypoints = [...new Set(graph.entrypoints)];
  for (const [index, entrypoint] of graph.entrypoints.entries()) {
    if (!nodesById.has(entrypoint)) {
      diagnostics.push(
        diagnostic("GE1008_MISSING_ENTRYPOINT", `Entrypoint '${entrypoint}' does not name a node`, {
          path: `#/entrypoints/${index}`,
          nodeIds: [entrypoint],
        }),
      );
    }
  }

  for (const [name, endpoint] of Object.entries(graph.outputs).sort(([left], [right]) =>
    compareUnicodeCodePoints(left, right),
  )) {
    if (!nodesById.has(endpoint.node)) {
      diagnostics.push(
        diagnostic("GE1009_MISSING_OUTPUT", `Output '${name}' references missing node '${endpoint.node}'`, {
          path: `#/outputs/${name}/node`,
          nodeIds: [endpoint.node],
        }),
      );
    }
  }

  // Identity and reference errors make adjacency ambiguous. Do not manufacture
  // cycle/reachability cascades from a graph whose structural index is invalid.
  if (diagnostics.length > 0) {
    return {
      valid: false,
      graphHash,
      canonicalGraph,
      entrypoints,
      topologicalLayers: [],
      diagnostics,
    };
  }

  const adjacency = new Map<string, string[]>();
  const inDegree = new Map<string, number>();
  const declarationIndex = new Map<string, number>();
  for (const [index, nodeId] of [...nodesById.keys()].entries()) {
    adjacency.set(nodeId, []);
    inDegree.set(nodeId, 0);
    declarationIndex.set(nodeId, index);
  }
  for (const edge of validEdges) {
    adjacency.get(edge.from.node)?.push(edge.to.node);
    inDegree.set(edge.to.node, (inDegree.get(edge.to.node) ?? 0) + 1);
  }
  const compareNodes = (left: string, right: string): number =>
    (declarationIndex.get(left) ?? Number.MAX_SAFE_INTEGER) -
      (declarationIndex.get(right) ?? Number.MAX_SAFE_INTEGER) ||
    compareUnicodeCodePoints(left, right);
  for (const targets of adjacency.values()) {
    targets.sort(compareNodes);
  }

  let currentLayer = [...nodesById.keys()].filter((nodeId) => inDegree.get(nodeId) === 0);
  const topologicalLayers: string[][] = [];
  const processed = new Set<string>();

  while (currentLayer.length > 0) {
    topologicalLayers.push(currentLayer);
    const nextLayer = new Set<string>();
    for (const nodeId of currentLayer) {
      processed.add(nodeId);
      for (const target of adjacency.get(nodeId) ?? []) {
        const nextDegree = (inDegree.get(target) ?? 0) - 1;
        inDegree.set(target, nextDegree);
        if (nextDegree === 0) {
          nextLayer.add(target);
        }
      }
    }
    currentLayer = [...nextLayer].sort(compareNodes);
  }

  const cycleNodes = [...nodesById.keys()].filter((nodeId) => !processed.has(nodeId));
  if (cycleNodes.length > 0) {
    diagnostics.push(
      diagnostic("GE1005_CYCLE", `Graph contains a cycle involving: ${cycleNodes.join(", ")}`, {
        nodeIds: cycleNodes,
      }),
    );

    return {
      valid: false,
      graphHash,
      canonicalGraph,
      entrypoints,
      topologicalLayers,
      diagnostics,
    };
  }

  for (const [index, entrypoint] of graph.entrypoints.entries()) {
    if (validEdges.some((edge) => edge.to.node === entrypoint)) {
      diagnostics.push(
        diagnostic(
          "GE1010_ENTRYPOINT_HAS_INCOMING",
          `Entrypoint '${entrypoint}' cannot have incoming edges`,
          { path: `#/entrypoints/${index}`, nodeIds: [entrypoint] },
        ),
      );
    }
  }

  // Reachability is deliberately rooted in explicit IR entrypoints. This keeps
  // disconnected parallel roots legal only when the author lists each one.
  const reachable = new Set<string>();
  const queue = entrypoints.filter((nodeId) => nodesById.has(nodeId));
  for (let index = 0; index < queue.length; index += 1) {
    const nodeId = queue[index] as string;
    if (reachable.has(nodeId)) {
      continue;
    }
    reachable.add(nodeId);
    for (const target of adjacency.get(nodeId) ?? []) {
      if (!reachable.has(target)) {
        queue.push(target);
      }
    }
  }
  for (const nodeId of [...nodesById.keys()].filter((id) => !reachable.has(id))) {
    diagnostics.push(
      diagnostic("GE1006_UNREACHABLE_NODE", `Node '${nodeId}' is not reachable from an entrypoint`, {
        nodeIds: [nodeId],
      }),
    );
  }

  const maxFanOut = graph.policies?.maxFanOut;
  if (typeof maxFanOut === "number") {
    for (const nodeId of nodesById.keys()) {
      const count = adjacency.get(nodeId)?.length ?? 0;
      if (count > maxFanOut) {
        diagnostics.push(
          diagnostic("GE1101_MAX_FAN_OUT", `Node '${nodeId}' fan-out ${count} exceeds policy ${maxFanOut}`, {
            nodeIds: [nodeId],
          }),
        );
      }
    }
  }

  const maxDepth = graph.policies?.maxDepth;
  if (typeof maxDepth === "number" && cycleNodes.length === 0 && topologicalLayers.length > maxDepth) {
    diagnostics.push(
      diagnostic(
        "GE1102_MAX_DEPTH",
        `Graph depth ${topologicalLayers.length} exceeds policy ${maxDepth}`,
      ),
    );
  }

  return {
    valid: diagnostics.every((item) => item.severity !== "error"),
    graphHash,
    canonicalGraph,
    entrypoints,
    topologicalLayers,
    diagnostics,
  };
}

File: python/src/graph_engineering/compiler.py
"""Deterministic compiler and structural validation for Graph IR."""

from __future__ import annotations

import hashlib
from collections import Counter, deque
from collections.abc import Iterable, Mapping
from dataclasses import dataclass
from enum import StrEnum
from types import MappingProxyType
from typing import Any

from pydantic import ValidationError

from .canonical import canonical_json
from .models import EdgeSpec, GraphSpec, NodeSpec, capture_graph_model_document
from .portable_json import portable_json_snapshot


class DiagnosticCode(StrEnum):
    DUPLICATE_NODE = "GE1001_DUPLICATE_NODE"
    DUPLICATE_EDGE = "GE1002_DUPLICATE_EDGE"
    MISSING_SOURCE = "GE1003_MISSING_SOURCE"
    MISSING_TARGET = "GE1004_MISSING_TARGET"
    CYCLE = "GE1005_CYCLE"
    UNREACHABLE_NODE = "GE1006_UNREACHABLE_NODE"
    INVALID_GRAPH = "GE1007_INVALID_GRAPH"
    MISSING_ENTRYPOINT = "GE1008_MISSING_ENTRYPOINT"
    MISSING_OUTPUT = "GE1009_MISSING_OUTPUT"
    ENTRYPOINT_HAS_INCOMING = "GE1010_ENTRYPOINT_HAS_INCOMING"
    MAX_FAN_OUT = "GE1101_MAX_FAN_OUT"
    MAX_DEPTH = "GE1102_MAX_DEPTH"
    MISSING_SOURCE_PORT = "GE1201_MISSING_SOURCE_PORT"
    MISSING_TARGET_PORT = "GE1202_MISSING_TARGET_PORT"
    PORT_SCHEMA_MISMATCH = "GE1203_PORT_SCHEMA_MISMATCH"
    DUPLICATE_TARGET_BINDING = "GE1204_DUPLICATE_TARGET_BINDING"
    INVALID_PORT_SCHEMA = "GE1205_INVALID_PORT_SCHEMA"
    OUTPUT_SCHEMA_MISMATCH = "GE1206_OUTPUT_SCHEMA_MISMATCH"
    ENTRYPOINT_SCHEMA_MISMATCH = "GE1207_ENTRYPOINT_SCHEMA_MISMATCH"
    UNSUPPORTED_TYPED_EDGE_MODE = "GE1208_UNSUPPORTED_TYPED_EDGE_MODE"
    UNSUPPORTED_GRAPH_REVISION = "GE1301_UNSUPPORTED_GRAPH_REVISION"
    GRAPH_IDENTITY_MISMATCH = "GE1302_GRAPH_IDENTITY_MISMATCH"
    COMPONENT_IDENTITY_MISMATCH = "GE1303_COMPONENT_IDENTITY_MISMATCH"
    INVALID_ROUTER_POLICY = "GE1401_INVALID_ROUTER_POLICY"
    UNSUPPORTED_EDGE_CONDITION = "GE1402_UNSUPPORTED_EDGE_CONDITION"
    CONDITION_SOURCE_NOT_ROUTER = "GE1403_CONDITION_SOURCE_NOT_ROUTER"
    ROUTE_NOT_ALLOWED = "GE1404_ROUTE_NOT_ALLOWED"
    DUPLICATE_ROUTE_CASE = "GE1405_DUPLICATE_ROUTE_CASE"
    DUPLICATE_ROUTE_TARGET = "GE1406_DUPLICATE_ROUTE_TARGET"
    INCOMPLETE_ROUTE_COVERAGE = "GE1407_INCOMPLETE_ROUTE_COVERAGE"
    INVALID_BARRIER_POLICY = "GE1421_INVALID_BARRIER_POLICY"
    BARRIER_POLICY_KIND_MISMATCH = "GE1422_BARRIER_POLICY_KIND_MISMATCH"
    BARRIER_NO_INPUTS = "GE1423_BARRIER_NO_INPUTS"
    BARRIER_THRESHOLD_EXCEEDS_INPUTS = "GE1424_BARRIER_THRESHOLD_EXCEEDS_INPUTS"


@dataclass(frozen=True, slots=True)
class Diagnostic:
    code: DiagnosticCode
    message: str
    node_id: str | None = None
    edge_id: str | None = None
    output_name: str | None = None
    path: str | None = None
    node_ids: tuple[str, ...] | None = None


@dataclass(frozen=True, slots=True)
class CompiledGraph:
    spec: GraphSpec
    canonical_graph: str
    graph_hash: str
    topological_layers: tuple[tuple[str, ...], ...]
    topological_order: tuple[str, ...]
    nodes: Mapping[str, NodeSpec]
    incoming: Mapping[str, tuple[EdgeSpec, ...]]
    outgoing: Mapping[str, tuple[EdgeSpec, ...]]


@dataclass(frozen=True, slots=True)
class CompilationResult:
    graph: CompiledGraph | None
    diagnostics: tuple[Diagnostic, ...]
    canonical_graph: str | None = None
    graph_hash: str | None = None
    entrypoints: tuple[str, ...] = ()
    topological_layers: tuple[tuple[str, ...], ...] = ()

    @property
    def valid(self) -> bool:
        return self.graph is not None and not self.diagnostics

    def raise_for_errors(self) -> CompiledGraph:
        if self.graph is None:
            raise GraphCompileError(self.diagnostics)
        return self.graph


class GraphCompileError(ValueError):
    def __init__(self, diagnostics: Iterable[Diagnostic]) -> None:
        self.diagnostics = tuple(diagnostics)
        summary = "; ".join(f"{item.code}: {item.message}" for item in self.diagnostics)
        super().__init__(summary or "graph compilation failed")


def _duplicate_values(values: Iterable[str]) -> list[str]:
    counts = Counter(values)
    return [value for value, count in counts.items() if count > 1]


def _invalid_result(exc: ValidationError) -> CompilationResult:
    details = "; ".join(
        f"{'.'.join(str(part) for part in error['loc'])}: {error['msg']}" for error in exc.errors()
    )
    return CompilationResult(
        graph=None,
        diagnostics=(Diagnostic(DiagnosticCode.INVALID_GRAPH, details),),
    )


def _unsafe_input_result() -> CompilationResult:
    return CompilationResult(
        graph=None,
        diagnostics=(
            Diagnostic(
                DiagnosticCode.INVALID_GRAPH,
                "graph input could not be inspected safely",
            ),
        ),
    )


def try_compile_graph(graph: GraphSpec | Mapping[str, Any]) -> CompilationResult:
    """Validate and compile a graph without raising for expected diagnostics."""

    try:
        document: object
        if type(graph) is GraphSpec:
            document = capture_graph_model_document(graph, GraphSpec)
        elif type(graph) is dict:
            document = portable_json_snapshot(graph)
        else:
            return _unsafe_input_result()
        validated = GraphSpec.model_validate(portable_json_snapshot(document))
        graph_document = portable_json_snapshot(capture_graph_model_document(validated, GraphSpec))
        graph = GraphSpec.model_validate(graph_document)
    except ValidationError as exc:
        return _invalid_result(exc)
    except Exception:
        return _unsafe_input_result()

    try:
        canonical_graph = canonical_json(graph_document)
        graph_hash = hashlib.sha256(canonical_graph.encode("utf-8")).hexdigest()
    except Exception:
        return _unsafe_input_result()
    entrypoints = tuple(graph.entrypoints)

    diagnostics: list[Diagnostic] = []
    duplicate_nodes = _duplicate_values(node.id for node in graph.nodes)
    duplicate_edges = _duplicate_values(edge.id for edge in graph.edges)

    for node_id in duplicate_nodes:
        diagnostics.append(
            Diagnostic(
                DiagnosticCode.DUPLICATE_NODE,
                f"node id {node_id!r} is declared more than once",
                node_id=node_id,
            )
        )
    for edge_id in duplicate_edges:
        diagnostics.append(
            Diagnostic(
                DiagnosticCode.DUPLICATE_EDGE,
                f"edge id {edge_id!r} is declared more than once",
                edge_id=edge_id,
            )
        )

    node_ids = {node.id for node in graph.nodes}
    for edge in graph.edges:
        if edge.source.node not in node_ids:
            diagnostics.append(
                Diagnostic(
                    DiagnosticCode.MISSING_SOURCE,
                    f"edge {edge.id!r} references missing source {edge.source.node!r}",
                    node_id=edge.source.node,
                    edge_id=edge.id,
                )
            )
        if edge.target.node not in node_ids:
            diagnostics.append(
                Diagnostic(
                    DiagnosticCode.MISSING_TARGET,
                    f"edge {edge.id!r} references missing target {edge.target.node!r}",
                    node_id=edge.target.node,
                    edge_id=edge.id,
                )
            )

    for entrypoint in graph.entrypoints:
        if entrypoint not in node_ids:
            diagnostics.append(
                Diagnostic(
                    DiagnosticCode.MISSING_ENTRYPOINT,
                    f"entrypoint {entrypoint!r} does not reference a node",
                    node_id=entrypoint,
                )
            )
    for output_name, endpoint in graph.outputs.items():
        if endpoint.node not in node_ids:
            diagnostics.append(
                Diagnostic(
                    DiagnosticCode.MISSING_OUTPUT,
                    f"output {output_name!r} references missing node {endpoint.node!r}",
                    node_id=endpoint.node,
                    output_name=output_name,
                )
            )

    # Duplicate identities or missing references make adjacency ambiguous. Avoid
    # cascading diagnostics whose meaning depends on a valid structural index.
    if diagnostics:
        return CompilationResult(
            graph=None,
            diagnostics=tuple(diagnostics),
            canonical_graph=canonical_graph,
            graph_hash=graph_hash,
            entrypoints=entrypoints,
        )

    node_order = {node.id: index for index, node in enumerate(graph.nodes)}
    node_by_id = {node.id: node for node in graph.nodes}
    incoming_lists: dict[str, list[EdgeSpec]] = {node.id: [] for node in graph.nodes}
    outgoing_lists: dict[str, list[EdgeSpec]] = {node.id: [] for node in graph.nodes}
    for edge in graph.edges:
        incoming_lists[edge.target.node].append(edge)
        outgoing_lists[edge.source.node].append(edge)

    indegree = {node_id: len(edges) for node_id, edges in incoming_lists.items()}
    current = [node.id for node in graph.nodes if indegree[node.id] == 0]
    layers: list[tuple[str, ...]] = []
    visited: list[str] = []
    while current:
        current.sort(key=node_order.__getitem__)
        layer = tuple(current)
        layers.append(layer)
        visited.extend(layer)
        next_nodes: set[str] = set()
        for node_id in layer:
            for edge in outgoing_lists[node_id]:
                target = edge.target.node
                indegree[target] -= 1
                if indegree[target] == 0:
                    next_nodes.add(target)
        current = list(next_nodes)

    if len(visited) != len(graph.nodes):
        cycle_nodes = tuple(node.id for node in graph.nodes if indegree[node.id] > 0)
        return CompilationResult(
            graph=None,
            diagnostics=(
                Diagnostic(
                    DiagnosticCode.CYCLE,
                    f"graph contains a cycle involving: {', '.join(cycle_nodes)}",
                ),
            ),
            canonical_graph=canonical_graph,
            graph_hash=graph_hash,
            entrypoints=entrypoints,
            topological_layers=tuple(layers),
        )

    for entrypoint in graph.entrypoints:
        if incoming_lists[entrypoint]:
            diagnostics.append(
                Diagnostic(
                    DiagnosticCode.ENTRYPOINT_HAS_INCOMING,
                    f"entrypoint {entrypoint!r} must not have incoming edges",
                    node_id=entrypoint,
                )
            )

    reachable: set[str] = set(graph.entrypoints)
    queue = deque(graph.entrypoints)
    while queue:
        node_id = queue.popleft()
        for edge in outgoing_lists[node_id]:
            target = edge.target.node
            if target not in reachable:
                reachable.add(target)
                queue.append(target)

    for node in graph.nodes:
        if node.id not in reachable:
            diagnostics.append(
                Diagnostic(
                    DiagnosticCode.UNREACHABLE_NODE,
                    f"node {node.id!r} is not reachable from any entrypoint",
                    node_id=node.id,
                )
            )

    if graph.policies and graph.policies.max_fan_out is not None:
        for node in graph.nodes:
            fan_out = len(outgoing_lists[node.id])
            if fan_out > graph.policies.max_fan_out:
                diagnostics.append(
                    Diagnostic(
                        DiagnosticCode.MAX_FAN_OUT,
                        (
                            f"node {node.id!r} has fan-out {fan_out}, exceeding policy limit "
                            f"{graph.policies.max_fan_out}"
                        ),
                        node_id=node.id,
                    )
                )

    if graph.policies and graph.policies.max_depth is not None:
        depth = len(layers)
        if depth > graph.policies.max_depth:
            diagnostics.append(
                Diagnostic(
                    DiagnosticCode.MAX_DEPTH,
                    f"graph depth {depth} exceeds policy limit {graph.policies.max_depth}",
                )
            )

    # Import lazily because the validators project through this module's stable
    # Diagnostic type. They do not compile or mutate graphs.
    from .integrated_barrier import validate_integrated_barrier_snapshot
    from .integrated_router import validate_integrated_router_snapshot
    from .typed_ports import validate_strict_typed_ports

    diagnostics.extend(validate_integrated_router_snapshot(graph))
    diagnostics.extend(validate_integrated_barrier_snapshot(graph))
    diagnostics.extend(validate_strict_typed_ports(graph))
    if diagnostics:
        return CompilationResult(
            graph=None,
            diagnostics=tuple(diagnostics),
            canonical_graph=canonical_graph,
            graph_hash=graph_hash,
            entrypoints=entrypoints,
            topological_layers=tuple(layers),
        )

    incoming = MappingProxyType(
        {node_id: tuple(edges) for node_id, edges in incoming_lists.items()}
    )
    outgoing = MappingProxyType(
        {node_id: tuple(edges) for node_id, edges in outgoing_lists.items()}
    )
    compiled = CompiledGraph(
        spec=graph,
        canonical_graph=canonical_graph,
        graph_hash=graph_hash,
        topological_layers=tuple(layers),
        topological_order=tuple(visited),
        nodes=MappingProxyType(node_by_id),
        incoming=incoming,
        outgoing=outgoing,
    )
    return CompilationResult(
        graph=compiled,
        diagnostics=(),
        canonical_graph=canonical_graph,
        graph_hash=graph_hash,
        entrypoints=entrypoints,
        topological_layers=tuple(layers),
    )


def compile_graph(graph: GraphSpec | Mapping[str, Any]) -> CompiledGraph:
    """Compile a Graph IR document or raise :class:`GraphCompileError`."""

    return try_compile_graph(graph).raise_for_errors()
